src/domain/entities/Questionnaire/QuestionnaireQuestion.ts

```diff
@@ -152,100 +152,102 @@ export class QuestionnaireQuestion {
+        const allQuestionIdsRequiringUpdate = _(
+            questionnaire.rules.flatMap(rule => {
+                if (
+                    rule.dataElementIds.includes(updatedQuestion.id) ||
+                    rule.teAttributeIds.includes(updatedQuestion.id)
+                ) {
+                    const actionUpdates = rule.actions.flatMap(
+                        action => action?.dataElement?.id || action.trackedEntityAttribute?.id
+                    );
+
+                    return actionUpdates;
+                }
+            })
+        )
+            .compact()
+            .filter(id => processedQuestions.find(q => q.id === id) === undefined)
+            .uniq()
+            .value();
+
+        const allQuestionsRequiringUpdate = sortedUpdatedQuestions.filter(question =>
+            allQuestionIdsRequiringUpdate.includes(question.id)
+        );
+        if (allQuestionsRequiringUpdate.length === 0) return sortedUpdatedQuestions;
+
+        //3. Recursively update all questions that need to be updated as a side effect of the current question update
+        const finalUpdatesWithSideEffects = allQuestionsRequiringUpdate.reduce(
+            (acc, questionRequiringUpdate) => {
+                const currentApplicableRules = getApplicableRules(
+                    questionRequiringUpdate,
+                    questionnaire.rules,
+                    acc
+                );
+                //4. Maintain a dependency graph to avoid infinite recursive calls,
+                // once a question has been processed, it should not be processed again
+                processedQuestions.push(questionRequiringUpdate);
+                const updates = this.updateQuestions(
+                    processedQuestions,
+                    acc,
+                    questionRequiringUpdate,
+                    currentApplicableRules,
+                    questionnaire
+                );
+                return updates;
+            },
+            sortedUpdatedQuestions
+        );
+        return finalUpdatesWithSideEffects;
     }
 
     private static updateQuestion(question: Question, rules: QuestionnaireRule[]): Question {
         const updatedIsVisible = this.isQuestionVisible(question, rules);
         const updatedErrors = this.getQuestionWarningsAndErrors(question, rules);
 
-        return {
-            ...question,
-            isVisible: updatedIsVisible,
-            errors: updatedErrors,
-        };
+        if (question.isVisible !== updatedIsVisible) {
+            //If a question is hidden, reset its value
+            return {
+                ...question,
+                isVisible: updatedIsVisible,
+                errors: updatedErrors,
+                value: undefined,
+            };
+        } else
+            return {
+                ...question,
+                isVisible: updatedIsVisible,
+                errors: updatedErrors,
+            };
```

Comment:
The return value is the same except for the `value: undefined`. This does not seem accidental repetition, to keep it DRY, move the condition inside.

Reply:
I have faced this issue before as well, typescript cannot resolve the type unless the condition is an if else. Is there any other way to resolve this? 
<img width="1728" alt="Screenshot 2024-11-14 at 3 42 24 PM" src="https://github.com/user-attachments/assets/27092bf2-3a6d-42fb-ad9f-a936e00b297f">

Reply:
> I have faced this issue before as well, typescript cannot resolve the type unless the condition is an if else. Is there any other way to resolve this? 

I see, things like this happen for union types, it's normal. Try this workaround:

`...(question.isVisible !== updatedIsVisible ? { value: undefined } : {})`